overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,130 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
 	return gadgetInfo.Connections, nil
 }
 
+// revisionsToKeep returns a map from name to revisions that need to be kept
+// for all current snaps in the system state.
+//
+// A downloaded file is only kept if any of the following are true:
+//    1. The revision is in SnapState.Sequence
+//    2. The revision is in saved in refresh-candidates
+//    3. The revision is pointed to by a download task in an ongoing change
+//
+// It is the caller's responsibility to lock state before calling this function.
+func revisionsToKeep(st *state.State) (map[string][]snap.Revision, error) {
+	snapStates, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// initialize refresh hints
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	}
+
+	snapToRevisions := map[string][]snap.Revision{}
+	keep := func(name string, rev snap.Revision) {
+		snapToRevisions[name] = append(snapToRevisions[name], rev)
+	}
+
+	// keep revisions in snap's sequence
+	for snapName, snapst := range snapStates {
+		for _, seq := range snapst.Sequence {
+			keep(snapName, seq.Revision)
+		}
+	}
+
+	// keep revisions in refresh hints
+	for snapName, hint := range refreshHints {
+		keep(snapName, hint.Revision())
+	}
+
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, _, err := snapSetupAndState(t)
+			if err != nil {
+				return nil, err
+			}
+			keep(snapsup.InstanceName(), snapsup.Revision())
+		}
+	}
+
+	return snapToRevisions, nil
+}
+
+// snapRevisionsToKeep returns revisions that need to be kept for the given snap.
+//
+// It is the caller's responsibility to lock state before calling this function.
+func snapRevisionsToKeep(st *state.State, snapName string) ([]snap.Revision, error) {
+	snapToRevisions, err := revisionsToKeep(st)
+	if err != nil {
+		return nil, err
+	}
+	return snapToRevisions[snapName], nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
+	snapToRevisions, err := revisionsToKeep(st)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for snapName, revisions := range snapToRevisions {
+		for _, revision := range revisions {
+			mountFile := snap.MountFile(snapName, revision)
+			if osutil.FileExists(mountFile) {
+				keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirState(dirs.SnapBlobDir, "*.snap", keepContent)
```

Comment:
unless you add some optimization to EnsureFilePath that does something reasonable if the target and the source have the same filename, this will end up I think comparing each snap file with itself which is very expensive. The Ensure* code was not written expecting the source and target to be the same. Simpler code might be doing  Glob and os.Remove of files not in the a map of the filenames to keep.